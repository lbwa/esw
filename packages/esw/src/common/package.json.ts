import path from 'path'
import fs from 'fs'
import { of, iif, map, throwError } from 'rxjs'
import { PackageJson } from 'type-fest'

export function resolvePackageJson(cwd: string) {
  const resolvedPkgJsonPath = path.resolve(cwd, 'package.json')
  const pkgJsonPath$ = of(resolvedPkgJsonPath)
  const resolve$ = pkgJsonPath$.pipe(
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    map(pkgJsonPath => require(pkgJsonPath) as PackageJson)
  )
  return iif(
    () => fs.existsSync(resolvedPkgJsonPath),
    resolve$,
    throwError(
      () =>
        new Error(
          `package.json file doesn't exists in the ${resolvedPkgJsonPath}`
        )
    )
  )
}
