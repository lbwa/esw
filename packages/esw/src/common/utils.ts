import path from 'path'
import { mkdirSync, writeFileSync } from 'fs'

// https://github.com/microsoft/TypeScript/pull/39094#Spreads_in_array_literals
export function tuple<T extends unknown[]>(args: [...T]) {
  return args
}

export function isProduction(process: NodeJS.Process) {
  return process.env['NODE_ENV'] === 'production'
}

export function isFulfillResult<Data>(
  result: PromiseSettledResult<Data>
): result is PromiseFulfilledResult<Data> {
  return result.status === 'fulfilled'
}

export async function writeToDiskSync(
  outPath: string,
  content: string | NodeJS.ArrayBufferView
) {
  mkdirSync(path.dirname(outPath), { recursive: true })
  writeFileSync(outPath, content, { encoding: null })
}
