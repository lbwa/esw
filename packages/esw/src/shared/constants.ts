export enum ProcessCode {
  OK,
  ERROR
}
