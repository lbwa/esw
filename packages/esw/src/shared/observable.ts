import { Observable, of, map } from 'rxjs'

export function lazyRequireObs<M = unknown>(filepath: string): Observable<M> {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return of(filepath).pipe(map(p => require(p)))
}
