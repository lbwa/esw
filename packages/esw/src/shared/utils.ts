export function isDef<V>(value: V): value is NonNullable<V> {
  return value !== null && value !== void 0
}

// https://github.com/microsoft/TypeScript/pull/39094#Spreads_in_array_literals
export function tuple<T extends unknown[]>(args: [...T]) {
  return args
}
