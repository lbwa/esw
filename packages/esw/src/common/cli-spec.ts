import { tuple } from './utils'

// shareable cli args spec
export const UNI_ARGS_SPEC = {
  // esbuild args
  '--absWorkingDir': String,
  '--allowOverwrite': Boolean,
  '--assetNames': String,
  // '--banner'
  '--bundle': Boolean,
  '--charset': function (val: string) {
    return val === 'ascii' ? val : null
  },
  '--chunkNames': String,
  '--color': Boolean,
  '--conditions': tuple([String]),
  // '--define'
  '--entryNames': String,
  '--entryPoints': tuple([String]),
  '--external': tuple([String]),
  // '--footer'
  '--format': function (val: string) {
    return ['cjs', 'esm', 'iife'].includes(val) ? val : null
  },
  '--globalName': String,
  '--incremental': Boolean,
  '--inject': tuple([String]),
  '--jsx': function (val: string) {
    return ['transform', 'preserve'].includes(val) ? val : null
  },
  '--jsxFactory': String,
  '--jsxFragment': String,
  '--keepNames': Boolean,
  '--legalComments': function (val: string) {
    return ['none', 'inline', 'eof', 'linked', 'external'].includes(val)
      ? val
      : null
  },
  // '--loader'
  '--logLevel': function (val: string) {
    return ['verbose', 'debug', 'info', 'warning', 'error', 'silent'].includes(
      val
    )
      ? val
      : null
  },
  '--logLimit': Number,
  '--mainFields': tuple([String]),
  '--metafile': Boolean,
  '--minify': Boolean,
  '--minifyIdentifiers': Boolean,
  '--minifySyntax': Boolean,
  '--minifyWhitespace': Boolean,
  '--nodePaths': tuple([String]),
  // '--outExtension'
  '--outbase': String,
  '--outdir': String,
  '--outfile': String,
  '--platform': (val: string) =>
    ['browser', 'node', 'neutral'].includes(val) ? val : null,
  // '--plugins'
  '--preserveSymlinks': Boolean,
  '--publicPath': String,
  '--pure': tuple([String]),
  '--resolveExtensions': tuple([String]),
  '--sourceRoot': String,
  '--sourcemap': function (val: string) {
    return ['true', 'false'].includes(val) ? val === 'true' : val
  },
  '--sourcesContent': Boolean,
  '--splitting': Boolean,
  // '--stdin'
  '--target': tuple([String]),
  '--treeShaking': function (val: string) {
    return val === 'true' || (val === 'ignore-annotations' ? val : null)
  },
  '--tsconfig': String,
  '--write': Boolean,

  // non build args
  '--help': Boolean,

  // alias
  '-h': '--help'
}
