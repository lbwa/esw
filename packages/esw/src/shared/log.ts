/* eslint-disable no-console */
import chalk from 'chalk'
import { isDef } from '@eswjs/common'
import { ProcessCode } from './constants'

const prefixes = {
  error: chalk.red('error') + ' -',
  info: chalk.cyan('info') + '  -'
}

export function error(...message: string[]) {
  console.error(prefixes.error, ...message)
}

export function info(...message: string[]) {
  console.log(prefixes.info, ...message)
}

export function printAndExit(
  message: string,
  code: number = ProcessCode.ERROR
) {
  if (code === ProcessCode.OK) {
    console.info(message)
  } else if (isDef(code)) {
    console.error(message)
  }
  process.exit(code)
}
