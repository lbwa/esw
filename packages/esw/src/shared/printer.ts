import { isDef } from '@eswjs/common'
import { ProcessCode } from './constants'

export function printToTerminal(
  message: string,
  code: number = ProcessCode.ERROR,
  exit = false
) {
  if (code === ProcessCode.OK) {
    process.stdout.write(message)
  } else if (isDef(code)) {
    process.stderr.write(message)
  }
  if (exit) {
    process.exit(code)
  }
}
