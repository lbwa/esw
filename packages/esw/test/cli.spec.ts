import fs from 'fs'
import path from 'path'
import type { SpawnSyncOptions } from 'child_process'
import spawn from 'cross-spawn'
import { ExitCode } from '@eswjs/common'
import { resolveFixture } from './shared'

const libDist = path.resolve(__dirname, '..', 'dist')
const binDist = path.resolve(libDist, 'bin')
const testRoot = path.resolve(__dirname, '.')
const eswBinary = path.resolve(binDist, 'esw.js')

beforeAll(async () => {
  const withDeps = path.resolve(testRoot, 'with-deps')
  await fs.promises.rm(path.resolve(withDeps, 'dist'), {
    force: true,
    recursive: true
  })
})

async function createBuildScript(
  fixturePath: string,
  commandArgs: string[] = ['build'],
  outFiles: Partial<Record<'esm' | 'cjs', string>> = {},
  spawnOptions = {} as SpawnSyncOptions
) {
  function resolveFixture(p: string) {
    return path.resolve(fixturePath, p)
  }
  const result = spawn.sync('node', [eswBinary, ...commandArgs], {
    encoding: 'utf-8',
    ...spawnOptions,
    cwd: fixturePath
  })
  // Process exited too early. The system ran out of memory or someone
  // called `kill -9` on the process
  expect(result.signal).not.toEqual('SIGKILL')
  // Process exited too early. The system could be shutting down or someone
  // might have called `kill` or `killall`
  expect(result.signal).not.toEqual('SIGTERM')
  expect(result.error).toBeNull()
  expect(result.status).toEqual(ExitCode.OK)

  process.stdout.write(result.output.filter(Boolean).join('\n'))

  Object.values(outFiles).forEach(p => {
    expect(fs.existsSync(resolveFixture(p))).toBeTruthy()
    expect(result.stdout).toContain(p)
  })
  const { esm, cjs } = outFiles
  return Promise.all(
    [esm, cjs].filter(Boolean).map(p =>
      fs.promises.readFile(
        path.resolve(resolveFixture(fixturePath), p as string),
        {
          encoding: 'utf8'
        }
      )
    )
  )
}

describe('esw cli', () => {
  it('should print root help message', () => {
    const shouldPrintHelp = spawn.sync('node', [eswBinary, '--help'], {
      cwd: process.cwd(),
      encoding: 'utf8'
    })
    expect(shouldPrintHelp.output).toMatchSnapshot(
      '--help should print help message'
    )

    const shouldPrintHelpWithAlias = spawn.sync('node', [eswBinary, '-h'], {
      cwd: process.cwd(),
      encoding: 'utf8'
    })
    expect(shouldPrintHelpWithAlias.output).toMatchSnapshot(
      '-h should print help message'
    )
    expect(shouldPrintHelpWithAlias.output).toEqual(shouldPrintHelp.output)

    const shouldPrintHelpWithWrongCommand = spawn.sync(
      'node',
      [eswBinary, 'wrongCommand'],
      {
        cwd: process.cwd(),
        encoding: 'utf8'
      }
    )

    expect(shouldPrintHelpWithWrongCommand.output).toMatchSnapshot(
      'wrong command should print help message'
    )
    expect(shouldPrintHelp.output).toEqual(
      shouldPrintHelpWithWrongCommand.output
    )
  })

  it('should print the esw version message', () => {
    const versionReg = /v\d+\.\d+\.\d+/

    ;[
      [eswBinary, '--version'],
      [eswBinary, '-v'],
      [eswBinary, 'build', '--version']
    ].forEach(args => {
      const shouldPrintVersionMsg = spawn.sync('node', args, {
        encoding: 'utf-8'
      })
      expect(shouldPrintVersionMsg.stdout).toMatch(versionReg)
    })
  })
})

describe('esw cli - build', () => {
  it('should work with no-options', async () => {
    await Promise.all(
      ['index.esm.js', 'index.js'].map(file =>
        fs.promises.rm(path.resolve(testRoot, 'fixture/no-options', file), {
          force: true,
          recursive: true
        })
      )
    )
    const fixturePath = resolveFixture('no-options')
    const [esmOutput, cjsOutput] = await createBuildScript(
      fixturePath,
      ['build'],
      { esm: 'index.esm.js', cjs: 'dist/cjs/index.js' }
    )

    expect(esmOutput).not.toContain('__esModule')
    expect(cjsOutput).toContain('__esModule')
  })

  it('should mark all peerDependencies and dependencies as external', async () => {
    const [esmOutput, cjsOutput] = await createBuildScript(
      resolveFixture('with-deps'),
      ['build'],
      {
        esm: 'dist/index.esm.js',
        cjs: 'dist/index.js'
      }
    )

    expect(esmOutput).toContain(`from "react"`)
    expect(esmOutput).toContain(`from "rxjs"`)
    expect(esmOutput).toContain(`from "rxjs/operators"`)

    expect(cjsOutput).toContain(`require("react")`)
    expect(cjsOutput).toContain(`require("rxjs")`)
    expect(cjsOutput).toContain(`require("rxjs/operators")`)
  })

  it('should print the help message of build command', () => {
    const shouldPrintHelpMsg = spawn.sync(
      'node',
      [eswBinary, 'build', '--help'],
      {
        cwd: process.cwd(),
        encoding: 'utf8'
      }
    )
    expect(shouldPrintHelpMsg.output).toMatchSnapshot(
      'build --help should print help message'
    )

    const shouldPrintHelpWithWrongCommand = spawn.sync(
      'node',
      [eswBinary, 'build', '--mini'],
      {
        cwd: process.cwd(),
        encoding: 'utf8'
      }
    )
    expect(shouldPrintHelpWithWrongCommand.stderr).toMatch(/error/i)
    expect(shouldPrintHelpWithWrongCommand.stderr).toContain(
      'Unknown arguments: --mini'
    )
    expect(shouldPrintHelpMsg.stderr).toEqual(
      shouldPrintHelpWithWrongCommand.stdout
    )

    const shouldPrintHelpWithAlias = spawn.sync(
      'node',
      [eswBinary, 'build', '-h'],
      {
        cwd: process.cwd(),
        encoding: 'utf8'
      }
    )
    expect(shouldPrintHelpWithAlias.output).toMatchSnapshot(
      'build -h should print help message'
    )
    expect(shouldPrintHelpWithAlias.stdout).toEqual(shouldPrintHelpMsg.stdout)
  })

  it('should work with uniq main and module field', async () => {
    await Promise.all(
      ['lib.js', 'index.js'].map(file =>
        fs.promises.rm(
          path.resolve(testRoot, 'fixture/uniq-main-module', file),
          {
            force: true,
            recursive: true
          }
        )
      )
    )
    const [esm, cjs] = await createBuildScript(
      path.resolve(testRoot, 'fixture/uniq-main-module'),
      ['build', 'index.ts'],
      {
        esm: `dist/lib.js`,
        cjs: `dist/index.js`
      }
    )

    expect(esm).not.toContain(`__esModule`)
    expect(cjs).toContain(`__esModule"`)
  })

  it('should throw a error message when building failed', async () => {
    const result = spawn.sync('node', [eswBinary, 'build'], {
      cwd: resolveFixture('build-error'),
      encoding: 'utf8'
    })
    expect(result.status).toEqual(ExitCode.ERROR)
    expect(result.stderr).toMatch(/could not resolve/i)
    expect(result.stdout).toEqual('')
  })
})
