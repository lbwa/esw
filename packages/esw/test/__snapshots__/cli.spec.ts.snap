// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`esw cli - build should print the help message of build command: build --help should print help message 1`] = `
Array [
  null,
  "
Description
  Compiles the codebase for publishing npm package.

Usage
  $ esw build [entry files]

  [entry file] represents the library entry point.
  If no entry is provided, the basename from main and module field in package.json will be used.
  User should always specific a entry point explicitly when the main and module have a different basename.

",
  "",
]
`;

exports[`esw cli - build should print the help message of build command: build -h should print help message 1`] = `
Array [
  null,
  "
Description
  Compiles the codebase for publishing npm package.

Usage
  $ esw build [entry files]

  [entry file] represents the library entry point.
  If no entry is provided, the basename from main and module field in package.json will be used.
  User should always specific a entry point explicitly when the main and module have a different basename.

",
  "",
]
`;

exports[`esw cli should print root help message: --help should print help message 1`] = `
Array [
  null,
  "
Usage
  $ esw <command>

Available commands
  build, watch

Options
  --version, -v      Show version number
  --help, -h         Display help messages

For more information run a command with the --help flag
  $ esw build --help

",
  "",
]
`;

exports[`esw cli should print root help message: -h should print help message 1`] = `
Array [
  null,
  "
Usage
  $ esw <command>

Available commands
  build, watch

Options
  --version, -v      Show version number
  --help, -h         Display help messages

For more information run a command with the --help flag
  $ esw build --help

",
  "",
]
`;

exports[`esw cli should print root help message: wrong command should print help message 1`] = `
Array [
  null,
  "
Usage
  $ esw <command>

Available commands
  build, watch

Options
  --version, -v      Show version number
  --help, -h         Display help messages

For more information run a command with the --help flag
  $ esw build --help

",
  "",
]
`;
